src/modules/timepix3/handlers/TimingEventHandler.h

```diff
@@ -10,112 +10,111 @@
+/// configuration file in the future.
+///
+/// \see Observer::DataEventObserver
+/// \see Observer::DataEventObservable
+/// \see timepixReadout::TDCReadout
+/// \see timepixReadout::EVRReadout
+/// \see timepixDTO::ESSGlobalTimeStamp
+///
+class TimingEventHandler
+    : public Observer::DataEventObserver<timepixReadout::TDCReadout>,
+      public Observer::DataEventObserver<timepixReadout::EVRReadout>,
+      public Observer::DataEventObservable<timepixDTO::ESSGlobalTimeStamp> {
+private:
+  /// \todo move this into configuration file
+  const milliseconds THRESHOLD_MS = milliseconds(20);
+
+  Counters &statCounters;
+
+  const nanoseconds FrequencyPeriodNs;
+
+  unique_ptr<timepixDTO::TDCDataEvent> lastTDCData;
+  unique_ptr<timepixDTO::EVRDataEvent> lastEVRData;
+
+  uint32_t tdcRepetitionFrequency{
+      static_cast<uint32_t>(FrequencyPeriodNs.count())};
+
   ///
-  /// \class TimingEventHandler
-  /// \brief Handles timing events for TDCReadout and EVRReadout data.
+  /// \brief Checks if the timing difference between the last TDCDataEvent and
+  /// EVRDataEvent is lower than the threshold.
   ///
-  /// This class is responsible for handling timing events related to TDCReadout
-  /// and EVRReadout data. It implements the Observer pattern as a
-  /// DataEventObserver for TDCReadout and EVRReadout, and as a
-  /// DataEventObservable for ESSGlobalTimeStamp.
+  /// \return true if the timing difference is lower than the threshold, false
+  /// otherwise.
   ///
-  /// The class maintains a threshold value for the timing difference between
-  /// TDC and EVR events, and provides a method to check if the last timing
-  /// difference is lower than the threshold.
+  inline bool isLastTimingDiffLowerThenThreshold() {
+    if (lastEVRData == nullptr || lastTDCData == nullptr) {
+      return false;
+    }
+
+    auto arrivalDiff =
+        abs(duration_cast<nanoseconds>(lastTDCData->arrivalTimestamp -
+                                       lastEVRData->arrivalTimestamp)
+                .count());
+
+    return arrivalDiff <= duration_cast<nanoseconds>(THRESHOLD_MS).count();
+  }
+
+public:
   ///
-  /// The class also stores the last TDCDataEvent and EVRDataEvent received, and
-  /// provides methods to apply data to these events.
+  /// \brief Constructs a TimingEventHandler object.
   ///
-  /// \note The threshold value is currently hard-coded and can be moved to a
-  /// configuration file in the future.
+  /// \param statCounters The reference to the Counters object for tracking
+  /// statistics.
   ///
-  /// \see Observer::DataEventObserver
-  /// \see Observer::DataEventObservable
-  /// \see timepixReadout::TDCReadout
-  /// \see timepixReadout::EVRReadout
-  /// \see timepixDTO::ESSGlobalTimeStamp
+  TimingEventHandler(Counters &statCounters, const int &Frequency = 14)
+      : statCounters(statCounters),
+        FrequencyPeriodNs(esstime::hzToNanoseconds(Frequency)) {}
```

Comment:
I suppose the `Frequency` is zero or positive. Would `size_t` or `uint` as type make more sense? Also, pass by value

```suggestion
  TimingEventHandler(Counters &statCounters, int Frequency = 14)
      : statCounters(statCounters),
        FrequencyPeriodNs(esstime::hzToNanoseconds(Frequency)) {}
```